### PR TITLE
OBC Firmware

### PR DESCRIPTION
**Pull request by chosterto**

### CODE REVIEW

File: firmware/teensy/teensy.ino [hunk 1]
```diff
@@ -55,12 +97,22 @@ void loop()
...
     {
         ICM.getAGMT();
-        imu_msg.linear_acceleration.x = ICM.accX() * MG_TO_MS2;
-        imu_msg.linear_acceleration.y = ICM.accY() * MG_TO_MS2;
-        imu_msg.linear_acceleration.z = ICM.accZ() * MG_TO_MS2;
+        updateIMU(&ICM);
     }
 
-    rcl_publish(&pub, &imu_msg, NULL);
+    if ( (millis() - prev_time1) > 5000)
```
[the remark this answers is not in the document]
Comment (haashimrehan) [reply firmware/teensy/teensy.ino:103]: or as we discussed, setAutoPVTcallbackPtr()

File: firmware/teensy/teensy.ino [hunk 2]
```diff
@@ -1,52 +1,94 @@
...
+
+void updateGNSS(SFE_UBLOX_GNSS* gnss)
+{
+    gps_msg.latitude = (double)( gnss->getLatitude() ) * 0.0000001;
+    gps_msg.longitude = (double)( gnss->getLongitude() ) * 0.0000001;
+    gps_msg.altitude = (double)( gnss->getAltitude() ) * 0.001;
+
+    double H_m = gnss->getHorizontalAccuracy() * 0.001;
+    double V_m = gnss->getVerticalAccuracy() * 0.001;
+    gps_msg.position_covariance[0] = H_m*H_m;
```
Comment (haashimrehan) [inline firmware/teensy/teensy.ino:62]: lets also set 
position_covariance_type = sensor_msgs::NavSatFix::COVARIANCE_TYPE_DIAGONAL_KNOWN;
in the [NavSatFix Message](http://docs.ros.org/en/noetic/api/sensor_msgs/html/msg/NavSatFix.html)

File: firmware/teensy/teensy.ino [hunk 3]
```diff
@@ -1,52 +1,94 @@
...
+    imu_msg.linear_acceleration.z = icm->accZ() * MG_TO_MS2;
+
+    imu_msg.angular_velocity.x = icm->gyrX() * DEG_TO_RAD;
+    imu_msg.angular_velocity.y = icm->gyrY() * DEG_TO_RAD;
+    imu_msg.angular_velocity.z = icm->gyrZ() * DEG_TO_RAD;
+}
+
+
+void updateGNSS(SFE_UBLOX_GNSS* gnss)
+{
```
Comment (haashimrehan) [inline firmware/teensy/teensy.ino:55]: lets also set the NavSatStatus 'status' component of the message here

here is an [example implementation](https://github.com/KumarRobotics/ublox/blob/4f107f3b82135160a1aca3ef0689fd119199bbef/ublox_gps/src/node.cpp#L786)

Comment (haashimrehan) [inline firmware/teensy/teensy.ino:55]: Remember to include the header of the message. We will need to include the following in it
- header.stamp
  - set it to ros::Time::now();
- header.frame_id
  - 'gps_frame'

**Comment by chosterto** [comment]
TODO:
Change Wire -> Wire1 and Serial1 -> Serial5

**Pull request merged**